dds/DCPS/XTypes/DynamicDataImpl.cpp

```diff
@@ -7,3263 +7,6479 @@
   }
   return good ? DDS::RETCODE_OK : DDS::RETCODE_ERROR;
 }
 
-DDS::ReturnCode_t DynamicDataImpl::get_string_value(ACE_CDR::Char*& value, MemberId id)
+DDS::ReturnCode_t DynamicDataImpl::set_string_value(DDS::MemberId id, const char* value)
 {
-  return get_single_value<TK_STRING8>(value, id);
+  return set_single_value<TK_STRING8>(id, value);
 }
 
-DDS::ReturnCode_t DynamicDataImpl::get_wstring_value(ACE_CDR::WChar*& value, MemberId id)
+DDS::ReturnCode_t DynamicDataImpl::set_wstring_value(DDS::MemberId id, const CORBA::WChar* value)
 {
 #ifdef DDS_HAS_WCHAR
-  return get_single_value<TK_STRING16>(value, id);
+  return set_single_value<TK_STRING16>(id, value);
 #else
   return DDS::RETCODE_UNSUPPORTED;
 #endif
 }
 
-DDS::ReturnCode_t DynamicDataImpl::get_complex_value(DDS::DynamicData_ptr& value, MemberId id)
+bool DynamicDataImpl::set_complex_to_struct(DDS::MemberId id, DDS::DynamicData_ptr value)
 {
-  ScopedChainManager chain_manager(*this);
-
-  const DDS::DynamicType_var base_type = get_base_type(type_);
-  DDS::TypeDescriptor_var descriptor;
-  if (base_type->get_descriptor(descriptor) != DDS::RETCODE_OK) {
+  DDS::DynamicTypeMember_var member;
+  if (type_->get_member(member, id) != DDS::RETCODE_OK) {
+    return false;
+  }
+  DDS::MemberDescriptor_var md;
+  if (member->get_descriptor(md) != DDS::RETCODE_OK) {
     return false;
   }
 
-  const TypeKind tk = base_type->get_kind();
-  bool good = true;
-
-  switch (tk) {
-  case TK_STRUCTURE:
-    {
-      DDS::DynamicTypeMember_var member;
-      const DDS::ReturnCode_t retcode = base_type->get_member(member, id);
-      if (retcode != DDS::RETCODE_OK) {
-        if (DCPS::DCPS_debug_level >= 1) {
-          ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) DynamicDataImpl::get_complex_value -")
-                     ACE_TEXT(" Failed to get DynamicTypeMember for member with ID %d\n"), id));
-        }
-        good = false;
-        break;
-      }
+  const DDS::DynamicType_var member_type = get_base_type(md->type());
+  const DDS::DynamicType_ptr value_type = value->type();
+  if (!member_type || !value_type || !member_type->equals(value_type)) {
+    return false;
+  }
+  return insert_complex(id, value);
+}
 
-      DDS::MemberDescriptor_var md;
-      if (member->get_descriptor(md) != DDS::RETCODE_OK) {
-        good = false;
-        break;
-      }
-      if (!skip_to_struct_member(md, id)) {
-        good = false;
-      } else {
-        DDS::DynamicType_ptr member_type = md->type();
-        if (!member_type) {
-          good = false;
-        } else {
-          value = new DynamicDataImpl(strm_, member_type);
-        }
-      }
-      break;
+bool DynamicDataImpl::set_complex_to_union(DDS::MemberId id, DDS::DynamicData_ptr value,
+                                           const DDS::TypeDescriptor_var& descriptor)
+{
+  if (id == DISCRIMINATOR_ID) {
+    DDS::DynamicType_var disc_type = get_base_type(descriptor->discriminator_type());
+    if (!disc_type->equals(value->type())) {
```

Comment:
Use a _var to manage the reference returned by `value->type()` -- are there other places the same change is needed?